### PR TITLE
String.split via fastpath instead of precompiled Pattern

### PR DESCRIPTION
**Pull request by amarkevich**
https://stackoverflow.com/questions/29393086/pattern-split-slower-than-string-split

benchmark:

`
public class SplitTest {

    private static final int CNT = 131072;
    private static String[] STR = new String[CNT];
    private static final String SEPARATOR = " ";
    static {
        for (int i = 0; i < CNT; ++i) {
            STR[i] = i + SEPARATOR + i;
        }
    }

    @org.junit.Test
    public void stringSplit() {
        metrics("String.split", s -> s.split(SEPARATOR));
    }

    @org.junit.Test
    public void pattern() {
        metrics("Pattern", java.util.regex.Pattern.compile(SEPARATOR)::split);
    }

    private static void metrics(String label, Function<String, String[]> f) {
        Runtime.getRuntime().gc();
        long m = Runtime.getRuntime().freeMemory();
        long t = System.currentTimeMillis();
        for (int i = 0; i < CNT; ++i) {
            org.junit.Assert.assertSame(2, f.apply(STR[i]).length);
        }
        System.out.println(label + ": "
                + ((m - Runtime.getRuntime().freeMemory()) / 1000000) + " MB "
                + (System.currentTimeMillis() - t) + " ms");
    }

    public static void main(String[] args) {
        SplitTest splitTest = new SplitTest();
        for (int i = 0; i < 10; ++i) {
            splitTest.stringSplit();
            splitTest.pattern();
        }
    }
}
`

results using java version "1.8.0_192":
`
String.split: 33 MB 16 ms
Pattern: 57 MB 47 ms
`

### CODE REVIEW

**Comment by andymc12** [comment]
Changes look good - thanks again!

File: core/src/main/java/org/apache/cxf/common/util/StringUtils.java [hunk 1]
```diff
@@ -26,37 +26,33 @@
...
     }
 
+    /* String.split via fastpath preferred */
     public static String[] split(String s, String regex) {
         return split(s, regex, 0);
     }
+    @Deprecated
     public static String[] split(String s, String regex, int limit) {
-        Pattern p = PATTERN_MAP.getOrDefault(regex, Pattern.compile(regex));
+        Pattern p = PATTERN_MAP.computeIfAbsent(regex, key -> Pattern.compile(key));
```
Comment (reta) [inline core/src/main/java/org/apache/cxf/common/util/StringUtils.java]: There is an well-known issue with JDK8 related to `computeIfAbsent`, https://bugs.openjdk.java.net/browse/JDK-8161372, not sure if we should take into account or not ...

**Comment by coheigea** [comment]
Hi @amarkevich,

The following methods are not used in source (main) code any more and could be removed:

StringUtils.getFirstNotEmpty
StringUtils.splitAsStream

StringUtils.getParts could also be removed, it is called here:

AbstractHttpDestination: StringUtils.getParts(credentials, " ");
PackageUtils: StringUtils.getParts(getPackageName(cls), "\\.")

For both these cases you could replace with String.split + convert the non-empty values to a List.

Finally, that leaves HttpHeadersImpl as the only place that is calling StringUtils.split. For several of the cases here it is a single character, for others not. You could move the StringUtils split functionality into this class + see if you can re-use split for the case that it's a single character, otherwise use a pre-compiled pattern separator?

**Comment by amarkevich** [comment]
I'll prepare a new PR according to proposals

**Pull request merged**